odl/trafos/backends/pywt_bindings.py

```diff
@@ -0,0 +1,523 @@
+        the smallest possible number of decomposition coefficients.
+        Only available with ``impl='pywt'``, See `pywt.MODES.modes`.
+
+    Returns
+    -------
+    shape_list : list
+        List containing the shapes of the wavelet (approximation
+        and detail) coefficients at different scaling levels.
+
+        ``shape_list[0]`` = shape of approximation coefficients at
+        the coarsest level,
+
+        ``shape_list[1]`` = shape of the detailed coefficients at
+        the coarsest level,
+
+        ``shape_list[N]`` = shape of the detailed coefficients at
+        the finest level,
+
+        ``shape_list[N+1]`` = shape of the original image,
+
+        ``N`` =  number of scaling levels
+    """
+    if len(shape) not in (1, 2, 3):
+        raise ValueError('shape must have length 1, 2 or 3, got {}'
+                         ''.format(len(shape)))
+
+    wbasis = pywt_wbasis(wbasis)
+    pywt_mode = PAD_MODES_ODL2PYWT[pad_mode]
+
+    max_level = pywt.dwt_max_level(shape[0], filter_len=wbasis.dec_len)
+    if nscales > max_level:
+        raise ValueError('too many scaling levels, got {}, maximum useful '
+                         'level is {}'
+                         ''.format(nscales, max_level))
+
+    # dwt_coeff_len calculates the number of coefficients at the next
+    # scaling level given the input shape, the length of the filter and
+    # the applied mode.
+    # We use this in the following way (per dimension):
+    # - length[0] = original data length
+    # - length[n+1] = dwt_coeff_len(length[n], ...)
+    # - until n = nscales
+    shape_list = [shape]
+    for scale in range(nscales):
+        shp = tuple(pywt.dwt_coeff_len(n, filter_len=wbasis.dec_len,
+                                       mode=pywt_mode)
+                    for n in shape_list[scale])
+        shape_list.append(shp)
+
+    # Add a duplicate of the last entry for the approximation coefficients
+    shape_list.append(shape_list[-1])
+
+    # We created the list in reversed order compared to what pywt expects
+    shape_list.reverse()
+    return shape_list
+
+
+def pywt_coeff_to_array(coeff, shape_list):
+    """Convert a Pywavelets coefficient list into a flat array.
+
+    Related to 1D, 2D and 3D multilevel discrete wavelet transforms.
+
+    Parameters
+    ----------
+    coeff : ordered `sequence`
+        Sequence of coefficients organized in the following way:
+
+        In 1D:
+
+        ``[aN, (dN), ..., (d1)]``
+
+        In 2D:
+
+        ``[aaN, (adN, daN, ddN), ..., (ad1, da1, dd1)]``
+
+        In 3D:
+
+        ``[aaaN, (aadN, adaN, addN, daaN, dadN, ddaN, dddN), ...
+        (aad1, ada1, add1, daa1, dad1, dda1, ddd1)]``
+
+        The abbreviations are "a" for "approximation" and "d" for "detail".
+        The letter position encodes the coefficient type in the respective
+        axis. ``N`` stands for the number of scaling levels.
+
+    shape_list : `sequence`
```

Comment:
shouldnt this more appropriately be called `shapes` or perhaps `wavelet_shapes` or something?